### PR TITLE
Added support to manage WDIO in a remote computer

### PR DESCRIPTION
**Pull request by leandromonjec**
# PR Details

Added support to manage WDIO in a remote computer

## Description

Implemented a REST API that allows to send WDIO commands to another computer. This has been complemented with an updated of the WDIO API provided by this library, so the calls to new REST API are transparent for the end user. Additionally, added environment configurability to allow working in a remote environment.

## Related Issue

#82

## Motivation and Context

There's the need in some projects to remotely control several GUI applications (developed in Tauri) as part of the same WDIO test.

## How Has This Been Tested

* Added REST API has been tested with a Postman collection that exercises the new endpoints.
* Added local test suite to the repository to allow starting the REST API and doing tests over the systelab-components showcase.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation 
- [x] I have updated the documentation accordingly (README.md for each widget)
- [x] I have added tests to cover my changes (at least 1 spec for each widget with the same coverage as the master branch)
- [x] All tests (new and existing) passed
- [ ] A new branch needs to be created from master to evolve previous versions
- [x] Increase version in package.json following [Semantic Versioning](https://semver.org/)
- [x] Add the issue into the right [project](https://github.com/systelab/systelab-components-wdio-test/projects) with the proper status (In progress)

### CODE REVIEW

File: src/remote/server/application.api.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      return ErrorHandlerAPI.handle(res, err);
+    }
+  }
+
+  public static async getName(req: Request, res: Response): Promise<any> {
+    try {
+      AutomationEnvironment.setApplication(+req.params.id);
+      return res.status(HttpStatus.OK).json({name: Browser.getName()}).send();
+    } catch (err) {
+      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({"error": err}).send();
```
Comment (jperez1-via) [inline src/remote/server/application.api.ts:48]: Could it be used the generic ErrorHandlerAPI class in here also for the error handling? The same comment applies to the methods below.

File: src/remote/server/tauri-driver.api.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import { HttpStatus } from './http-status';
+
+
+export class TauriDriverAPI {
+    public static async start(req: Request, res: Response): Promise<any> {
+        try {
+            TauriDriverUtility.start();
+            return res.status(HttpStatus.NO_CONTENT).send();
+        } catch (err) {
+            return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({"error": err}).send();
```
Comment (jperez1-via) [inline src/remote/server/tauri-driver.api.ts:12]: Could it be used the generic ErrorHandlerAPI class in here also for the error handling? The same comment applies to the method below.

**Review by jperez1-via — APPROVED** [review]

File: remote-application-jasmine.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
+  "spec_dir": "test/remote",
```
Comment (joaquimvila) [inline remote-application-jasmine.json:2]: As this file is just for the sake of testing WDIO remote, I think it should be placed under the `test/remote` folder

File: remote-application-jasmine.setup.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import {SpecReporter, StacktraceOption} from 'jasmine-spec-reporter';
+import {StandaloneTestCaseReporter} from "./src";
```
Comment (joaquimvila) [inline remote-application-jasmine.setup.ts:2]: As this file is just for the sake of testing WDIO remote, I think it should be placed under the test/remote folder

File: src/remote/client/element-finder-remote.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  public async tap(): Promise<void> {
+    await this.executeEndpoint('POST', 'element/action/tap', {locators: this.locators});
+  }
+
+
+  // Condition waits
+  public async waitToBePresent(timeout: number = 500): Promise<void> {
+    await this.executeEndpoint('GET', 'element/wait/present', {locators: this.locators});
```
Comment (joaquimvila) [inline src/remote/client/element-finder-remote.ts:134]: This wait GET call should be updated to be a POST call. As we need to pass a request body (i.e locators), the HTTP protocol does not support so for GET method. As we do in other cases, this should be changed to be a POST.
This applies to all condition waits.

File: src/remote/server/application.api.ts (hunk 1 above)
Comment (joaquimvila) [reply src/remote/server/application.api.ts:48]: Absolutely, that one should be used here and in any other case.

File: src/remote/server/application.api.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      return res.status(HttpStatus.NO_CONTENT).send();
+    } catch (err) {
+      return ErrorHandlerAPI.handle(res, err);
+    }
+  }
+
+  public static async getName(req: Request, res: Response): Promise<any> {
+    try {
+      AutomationEnvironment.setApplication(+req.params.id);
+      return res.status(HttpStatus.OK).json({name: Browser.getName()}).send();
```
Comment (joaquimvila) [inline src/remote/server/application.api.ts:46]: I miss an `await` as now the Browser.getName() returns a promise.

Comment (joaquimvila) [reply src/remote/server/application.api.ts:46]: Same in following methods to gather general browser info.

File: test/remote/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+```json
+{
+  "url": "https://www.werfen.com"
+}
+```
+
+## Query a text
+
+Send a request to POST localhost:3333/wdio/applications/1/navigate endpoint with the following body:
```
Comment (joaquimvila) [inline test/remote/README.md:42]: The endpoint route is not correct

File: test/remote/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      "selector":"h1.info-container__title"
+    }
+  ]
+}
+```
+
+
+## Query a text
+
+Send a request to POST localhost:3333/wdio/applications/1/element/action/click endpoint with the following body:
```
Comment (joaquimvila) [inline test/remote/README.md:66]: This is not a query of a text. Maybe a click?

File: test/remote/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      "type":"ElementSelector",
+      "selector":"h1.info-container__title"
+    }
+  ]
+}
+```
+
+## Postman collection
+
+You can find more examples by importing the PostmanCollection.json on Postman. The examples can be found on WDIO Remote Server
```
Comment (joaquimvila) [inline test/remote/README.md:89]: Mark the JSON as a file by adding quotes surrounding it.

File: test/remote/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+
```
Comment (joaquimvila) [inline test/remote/README.md:91]: I miss a section explaining how to run the test suite added in this PR.

File: test/remote/remote-app.spec.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
+import {AutomationEnvironment, Browser, BrowserType, RemoteApplication, RemoteApplicationManager} from '../../src';
+import {ComponentsPage} from '../pages/components.po';
+
+describe('Remote application test', () => {
+  // Manually start the remote-application-server ( npm run start-remote-test-server) in a separate terminal
```
Comment (joaquimvila) [inline test/remote/remote-app.spec.ts:5]: This could be removed if added to previous README.md

File: test/remote/remote-scenario.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+// Server is started manually with the server.ts script
```
Comment (joaquimvila) [inline test/remote/remote-scenario.ts:1]: This was just an idea of how to structure the test. We should get rid of this file.

**Review by joaquimvila — APPROVED** [review]

File: src/remote/client/browser-remote.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  public static async getOperatingSystem(): Promise<string> {
+    const response = await this.executeEndpoint('GET', 'os');
+    const body = await response.json();
+    return body.os;
+  }
+
+
+  // Window
+  public static async getWindowSize(): Promise<{ width: number, height: number }> {
+    const response = await this.executeEndpoint('GET', 'os');
```
[the remark this answers is not in the document]
Comment (leandromonjec) [reply src/remote/client/browser-remote.ts:88]: Fixed! Good catch!

File: src/remote/server/application.api.ts (hunk 1 above)
Comment (leandromonjec) [reply src/remote/server/application.api.ts:48]: Applied suggestion

File: src/remote/server/tauri-driver.api.ts (hunk 2 above)
Comment (leandromonjec) [reply src/remote/server/tauri-driver.api.ts:12]: Applied

File: remote-application-jasmine.json (hunk 3 above)
Comment (leandromonjec) [reply remote-application-jasmine.json:2]: Applied

File: remote-application-jasmine.setup.ts (hunk 4 above)
Comment (leandromonjec) [reply remote-application-jasmine.setup.ts:2]: Moved

File: src/remote/client/element-finder-remote.ts (hunk 5 above)
Comment (leandromonjec) [reply src/remote/client/element-finder-remote.ts:134]: Fixed

File: test/remote/README.md (hunk 10 above)
Comment (leandromonjec) [reply test/remote/README.md:91]: Added:

## Run test suite collection

To run the test suite, first start the remote server by running on a terminal:

```shell
npm run start-remote-test-server
```
Then, on other terminal, run:
```shell
npm run test-remote-server
```

File: test/remote/remote-scenario.ts (hunk 12 above)
Comment (leandromonjec) [reply test/remote/remote-scenario.ts:1]: Removed

File: test/remote/remote-app.spec.ts (hunk 11 above)
Comment (leandromonjec) [reply test/remote/remote-app.spec.ts:5]: Removed

File: test/remote/README.md (hunk 7 above)
Comment (leandromonjec) [reply test/remote/README.md:42]: Endpoint route updated to `localhost:3333/wdio/applications/1/element/query/text`

**Pull request merged**